river/Cursor.zig

```diff
@@ -241,6 +280,68 @@ listen_motion_absolute: c.wl_listener = undefined,
 listen_motion: c.wl_listener = undefined,
 listen_request_set_cursor: c.wl_listener = undefined,
 
+pub fn createNewPointerConstraint(self: *Self, constraint: ?*c.wlr_pointer_constraint_v1) void {
+    const pointer_constraint = util.gpa.create(struct_pointer_constraint) catch {
+        return;
+    };
+    pointer_constraint.cursor = self;
+    pointer_constraint.constraint = constraint;
+
+    pointer_constraint.set_region.notify = handlePointerConstraintSetRegion;
+    c.wl_signal_add(&constraint.?.events.set_region, &pointer_constraint.set_region);
+
+    pointer_constraint.destroy.notify = handlePointerConstraintDestroy;
+    c.wl_signal_add(&constraint.?.events.destroy, &pointer_constraint.destroy);
+
+    const focus = &self.seat.focused;
+    if (focus.* == .view and focus.view.wlr_surface != null) {
+        if (focus.view.wlr_surface == constraint.?.surface) {
+            if (self.active_constraint == constraint) {
+                return;
+            }
+
+            c.wl_list_remove(&self.listen_constraint_commit.link);
+            if (self.active_constraint != null) {
+                if (constraint == null) {
+                    self.warpToConstraintCursorHint();
+                }
+                c.wlr_pointer_constraint_v1_send_deactivated(
+                    self.active_constraint);
+            }
+
+            self.active_constraint = constraint;
+
+            if (constraint == null) {
+                c.wl_list_init(&self.listen_constraint_commit.link);
+                return;
+            }
+
+            self.active_confine_requires_warp = true;
+
+            // FIXME: Big hack, stolen from wlr_pointer_constraints_v1.c:121.
+            // This is necessary because the focus may be set before the surface
+            // has finished committing, which means that warping won't work properly,
+            // since this code will be run *after* the focus has been set.
+            // That is why we duplicate the code here.
+            if (c.pixman_region32_not_empty(&constraint.?.current.region) > 0) {
+                const tst = c.pixman_region32_intersect(&constraint.?.region,
+                    &constraint.?.surface.*.input_region, &constraint.?.current.region);
+            } else {
+                const tst = c.pixman_region32_copy(&constraint.?.region,
+                    &constraint.?.surface.*.input_region);
+            }
```

Comment:
Yeah. I forgot to mention that most of the code is copied from sway. I added it just in case it is necessary (like the comment from sway's code says), but I'll look to see if there is a way around it.

Reply:
I am not sure 1:1 copying and translating what sway does is the right thing to do here. While GPL can technically swallow MIT, you'd have to include the copyright notice of everyone who ever worked on that part of sway, explicitly only for that part of the code, or you'd be violating sways license (I am not sure if `Copyright (c) the Sway authors` works, but `copied from sway` definitely does not).

It may be better to re-implement this feature, especially since the differences between river and sway are not just superficial but also concern the internal stuff.